[WebXR][OpenXR] Export OpenXR textures to WebProcess for WebGL rendering

svillar opened the pull request:
#### 02a244de37cdad18c9d2a3d3b8835e0a83d4a3ce
<pre>
[WebXR][OpenXR] Export OpenXR textures to WebProcess for WebGL rendering
<a href="https://bugs.webkit.org/show_bug.cgi?id=296617">https://bugs.webkit.org/show_bug.cgi?id=296617</a>

Reviewed by Carlos Garcia Campos.

OpenXR textures are created by OpenXR swapchains in the UIProcess. In order
to get the WebXR content rendered into the OpenXR compositor we need first
to pass the textures to the WebProcess so that WebGL content could be
rendered there. The most obvious match for that in Linux systems is using
DMABuf. We export the textures in the UIProcess and import them in the
WebProcess using the already existing ExternalTexture abstraction created
for Cocoa ports.

WebXROpaqueFramebuffer was updated so that it&apos;s an almost exact copy of
the Cocoa version (we should think about eventually unifying them back
again). The added code uses the GraphicsContext to import the OpenXR
textures into a render buffer which is then used by WebGL commands to
render the WebXR scene. We had to fix a bug added in 295361@main.
Basically it was forcing WebGL1 to use a NV extension for blitting
framebuffers. That was because that extension is available for Cocoa
ports but not for all linux machines. Instead that extension must be
used only if supported, otherwise the ANGLE blit should be chosen.

Thw WebXR fake device used for testing has been also updated as we use
the same data structures than Cocoa platforms from now on (there are
only a few small differences).

On the OpenXR side the main change is the addition of the code that
queries and selects the swapchain format to be used depending on what
is exactly supported by the platform.

Last but not least, some new tests pass now after this change. We had
to disable running WPT WebXR tests in debug mode because
WebXROpaqueFramebuffer is not prepared to work with the WebFakeXRDevice
yet. It tries to initialize some gfx stuff that depends on data coming
from the UI process (like the exported textures). That is not a problem
for release builds because we just bail out, but in debug builds
several assertions are not met.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/webxr/webGLCanvasContext_create_xrcompatible.https-expected.txt: Removed.
* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
(WebCore::ensure):
(WebCore::createAndBindCompositorBuffer):
(WebCore::makeExternalImageSource):
(WebCore::createAndBindTempBuffer):
(WebCore::toIntSize):
(WebCore::WebXROpaqueFramebuffer::~WebXROpaqueFramebuffer):
(WebCore::WebXROpaqueFramebuffer::startFrame):
(WebCore::WebXROpaqueFramebuffer::endFrame):
(WebCore::WebXROpaqueFramebuffer::releaseAllDisplayAttachments):
(WebCore::WebXROpaqueFramebuffer::resolveMSAAFramebuffer):
(WebCore::WebXROpaqueFramebuffer::blitShared):
(WebCore::WebXROpaqueFramebuffer::blitSharedToLayered):
(WebCore::WebXROpaqueFramebuffer::drawFramebufferSize const):
(WebCore::WebXROpaqueFramebuffer::drawViewport const):
(WebCore::displayLayout):
(WebCore::calcFramebufferPhysicalSize):
(WebCore::WebXROpaqueFramebuffer::setupFramebuffer):
(WebCore::WebXROpaqueFramebuffer::reusableDisplayAttachments const):
(WebCore::WebXROpaqueFramebuffer::bindCompositorTexturesForDisplay):
(WebCore::WebXROpaqueFramebuffer::reusableDisplayAttachmentsAtIndex):
(WebCore::WebXROpaqueFramebuffer::releaseDisplayAttachmentsAtIndex):
(WebCore::WebXROpaqueFramebuffer::bindResolveAttachments):
(WebCore::WebXROpaqueFramebuffer::calculateViewportShared): Deleted.
* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.h:
* Source/WebCore/platform/graphics/GraphicsContextGL.h:
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::blitFramebuffer):
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLTextureMapperGBM.cpp:
(WebCore::GraphicsContextGLTextureMapperGBM::createDrawingBuffer const):
(WebCore::GraphicsContextGLTextureMapperGBM::createExternalImage):
(WebCore::GraphicsContextGLTextureMapperGBM::bindExternalImage):
(WebCore::GraphicsContextGLTextureMapperGBM::enableRequiredWebXRExtensions):
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLTextureMapperGBM.h:
* Source/WebCore/platform/xr/PlatformXR.h:
* Source/WebCore/testing/WebFakeXRDevice.cpp:
(WebCore::SimulatedXRDevice::initializeTrackingAndRendering):
(WebCore::SimulatedXRDevice::shutDownTrackingAndRendering):
(WebCore::SimulatedXRDevice::frameTimerFired):
(WebCore::SimulatedXRDevice::createLayerProjection):
(WebCore::SimulatedXRDevice::deleteLayer):
* Source/WebCore/testing/WebFakeXRDevice.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Shared/XR/PlatformXR.serialization.in:
* Source/WebKit/UIProcess/XR/PlatformXRSystem.cpp:
(WebKit::PlatformXRSystem::invalidate):
* Source/WebKit/UIProcess/XR/openxr/OpenXRLayer.cpp:
(WebKit::OpenXRLayer::OpenXRLayer):
(WebKit::OpenXRLayer::exportOpenXRTexture):
(WebKit::OpenXRLayerProjection::create):
(WebKit::OpenXRLayerProjection::OpenXRLayerProjection):
(WebKit::OpenXRLayerProjection::startFrame):
* Source/WebKit/UIProcess/XR/openxr/OpenXRLayer.h:
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
(WebKit::OpenXRCoordinator::collectSwapchainFormatsIfNeeded):
(WebKit::OpenXRCoordinator::createSwapchain):
(WebKit::OpenXRCoordinator::createLayerProjection):
(WebKit::OpenXRCoordinator::scheduleAnimationFrame):
(WebKit::OpenXRCoordinator::populateFrameData):
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.h:

Canonical link: <a href="https://commits.webkit.org/298184@main">https://commits.webkit.org/298184@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/34c3f0f5d4c2c3ab5996e548486c6109b0666824

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114565 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34311 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24774 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120729 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65291 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/116454 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/34940 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/42871 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/87076 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/41984 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117513 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27843 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102882 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67467 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27028 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21009 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64411 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97221 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21122 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123936 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41578 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/31038 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/95896 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/41955 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99072 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95679 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24374 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40843 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18678 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/37648 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41457 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/46967 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41043 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44350 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42793 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebCore/platform/graphics/GraphicsContextGL.h:
```diff
@@ -83,6 +87,18 @@ using GraphicsContextGLExternalImageSource = Variant<
+    Vector<WTF::UnixFileDescriptor> fds;
+    Vector<uint32_t> strides;
+    Vector<uint32_t> offsets;
+    int fourcc;
```
We normally use uint32_t to fourcc too.

carlosgcampos commented on Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
```diff
@@ -2938,15 +2940,22 @@ void GraphicsContextGLANGLE::getActiveUniformBlockiv(PlatformGLObject program, G
+    if (target == RENDERBUFFER)
+        GL_EGLImageTargetRenderbufferStorageOES(RENDERBUFFER, eglImage);
+    else
+        GL_EGLImageTargetTexture2DOES(target, eglImage);
```
I think all these WEbXR methods should be implemented in GraphicsContextGLTextureMapperGBM

svillar replied:
Isn't it more flexible to have them here? It would allow platforms not using GBM to run that. Or is that combination not usually supported?

Also some of these methods could use GBM stuff to implement this texture sharing in a different way if I am not mistaken

svillar replied:
Clarified offline. The GBM implementation is the one that actually has the DMABuf requirements. Moving things there then..

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/GraphicsContextGLTextureMapperANGLE.cpp:
```diff
@@ -417,6 +426,63 @@ void GraphicsContextGLTextureMapperANGLE::enableFoveation(GCGLuint)
 {
 }
+
+GCGLExternalImage GraphicsContextGLTextureMapperANGLE::createExternalImage(ExternalImageSource&& source, GCGLenum, GCGLint)
```
This should be in GraphicsContextGLTextureMapperGBM too

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/GraphicsContextGLTextureMapperANGLE.cpp:
```diff
@@ -417,6 +426,63 @@ void GraphicsContextGLTextureMapperANGLE::enableFoveation(GCGLuint)
+    attributes.append(EGL_NONE);
+
+    auto display = m_displayObj;
+    if (display == EGL_NO_DISPLAY) {
```
Why do we need the display variable? can't we just use m_displayObj directly?

svillar replied:
Oh, just a leftover from the dark times when image importing was not working :)

carlosgcampos commented on Source/WebCore/platform/xr/PlatformXR.h:
```diff
@@ -281,33 +283,37 @@ struct FrameData {
+        Vector<WTF::UnixFileDescriptor> fds;
+        Vector<uint32_t> strides;
+        Vector<uint32_t> offsets;
+        int fourcc;
```
```suggestion
        uint32_t fourcc;
```

carlosgcampos commented on Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
```diff
@@ -8790,7 +8790,16 @@ using WebCore::GraphicsContextGL::ExternalImageSource = Variant<WebCore::Graphic
+    Vector<WTF::UnixFileDescriptor> fds;
+    Vector<uint32_t> strides;
+    Vector<uint32_t> offsets;
+    int fourcc;
```
```suggestion
    uint32_t fourcc;
```

carlosgcampos commented on Source/WebKit/Shared/XR/PlatformXR.serialization.in:
```diff
@@ -122,31 +121,36 @@ header: <WebCore/PlatformXR.h>
+    Vector<WTF::UnixFileDescriptor> fds;
+    Vector<uint32_t> strides;
+    Vector<uint32_t> offsets;
+    int fourcc;
```
```suggestion
    uint32_t fourcc;
```

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRLayer.cpp:
```diff
@@ -62,17 +72,93 @@ OpenXRLayerProjection::OpenXRLayerProjection(UniqueRef<OpenXRSwapchain>&& swapch
 }
 
-std::optional<PlatformXR::FrameData::LayerData> OpenXRLayerProjection::startFrame()
+std::optional<PlatformXR::FrameData::LayerData> OpenXRLayerProjection::startFrame(WebCore::PlatformDisplay& display, WebCore::GLContext& context)
```
You can pass the GLContext only and get the PlatformDisplay with GLContext::display()

svillar replied:
ah right, good suggestion

svillar replied:
Hmm, actually we can pass the context when creating the OpenXRLayer, it is not going to change between calls to startFrame so it does not seem correct to pass it every time

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
```diff
@@ -125,13 +125,41 @@ void OpenXRCoordinator::requestPermissionOnSessionFeatures(WebPageProxy& page, c
+    }
 
+    auto preferredFormat = alpha ? GL_RGBA8 : GL_RGB8;
+    auto format =  m_supportedSwapchainFormats.contains(preferredFormat) ? preferredFormat : m_supportedSwapchainFormats.first();
```
```suggestion
    auto format = m_supportedSwapchainFormats.contains(preferredFormat) ? preferredFormat : m_supportedSwapchainFormats.first();
```

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRLayer.cpp:
```diff
@@ -62,17 +72,93 @@ OpenXRLayerProjection::OpenXRLayerProjection(UniqueRef<OpenXRSwapchain>&& swapch
-    return PlatformXR::FrameData::LayerData {
-        .framebufferSize = m_swapchain->size(),
-        .opaqueTexture = *texture
+    auto addResult = m_exportedTextures.add(*texture, m_nextReusableTextureIndex);
```
I see we never remove entries from the map, can the swap chain delete textures?

svillar replied:
No it cannot. Different WebXR sessions would use different OpenXRLayers so once the session is finished we get rid of everything in the destructor

carlosgcampos reviewed (commented):
Looks great in general! I have only a few comments.

carlosgcampos commented on Source/WebCore/platform/graphics/gbm/GraphicsContextGLTextureMapperGBM.cpp:
```diff
@@ -242,6 +243,93 @@ void GraphicsContextGLTextureMapperGBM::prepareForDisplayWithFinishedSignal(Func
+    if (!makeContextCurrent())
+        return false;
+
+    return enableExtension("GL_OES_EGL_image"_s)
+        && enableExtension("GL_OES_EGL_image_external"_s)
+        && enableExtension("EGL_KHR_image_base"_s)
+        && enableExtension("EGL_EXT_image_dma_buf_import"_s)
+        && enableExtension("EGL_KHR_surfaceless_context"_s);
```
GL_OES_EGL_image is already enabled in platformInitializeExtensions, and maybe all others should be moved there too except GL_OES_EGL_image_external since they are needed for WebGL too not only for XR.

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRLayer.cpp:
```diff
@@ -30,34 +40,83 @@ namespace WebKit {
+    glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, m_swapchain->width(), m_swapchain->height(), 0, GL_RGBA, GL_UNSIGNED_BYTE, nullptr);
+
+    auto& display = m_glContext.display();
+    auto image = display.createEGLImage(m_glContext.platformContext(), EGL_GL_TEXTURE_2D, (EGLClientBuffer)(uint64_t)openxrTexture, { });
```
So, we keep a reference to the context just to get the display and platform context, but this is always called with the context made current, so we can avoid keeping the ref and just use GLContext::current()

carlosgcampos commented on Source/WebKit/UIProcess/XR/openxr/OpenXRLayer.cpp:
```diff
@@ -30,34 +40,88 @@ namespace WebKit {
+    if (!glContext) {
+        RELEASE_LOG(XR, "No current GL context to export OpenXR texture");
+        return std::nullopt;
+    }
```
I don't think this is expected tom happen, I would add an assert instead.

svillar committed:
[WebXR][OpenXR] Export OpenXR textures to WebProcess for WebGL rendering
https://bugs.webkit.org/show_bug.cgi?id=296617

Reviewed by Carlos Garcia Campos.

OpenXR textures are created by OpenXR swapchains in the UIProcess. In order
to get the WebXR content rendered into the OpenXR compositor we need first
to pass the textures to the WebProcess so that WebGL content could be
rendered there. The most obvious match for that in Linux systems is using
DMABuf. We export the textures in the UIProcess and import them in the
WebProcess using the already existing ExternalTexture abstraction created
for Cocoa ports.

WebXROpaqueFramebuffer was updated so that it's an almost exact copy of
the Cocoa version (we should think about eventually unifying them back
again). The added code uses the GraphicsContext to import the OpenXR
textures into a render buffer which is then used by WebGL commands to
render the WebXR scene. We had to fix a bug added in 295361@main.
Basically it was forcing WebGL1 to use a NV extension for blitting
framebuffers. That was because that extension is available for Cocoa
ports but not for all linux machines. Instead that extension must be
used only if supported, otherwise the ANGLE blit should be chosen.

Thw WebXR fake device used for testing has been also updated as we use
the same data structures than Cocoa platforms from now on (there are
only a few small differences).

On the OpenXR side the main change is the addition of the code that
queries and selects the swapchain format to be used depending on what
is exactly supported by the platform.

Last but not least, some new tests pass now after this change. We had
to disable running WPT WebXR tests in debug mode because
WebXROpaqueFramebuffer is not prepared to work with the WebFakeXRDevice
yet. It tries to initialize some gfx stuff that depends on data coming
from the UI process (like the exported textures). That is not a problem
for release builds because we just bail out, but in debug builds
several assertions are not met.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/webxr/webGLCanvasContext_create_xrcompatible.https-expected.txt: Removed.
* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
(WebCore::ensure):
(WebCore::createAndBindCompositorBuffer):
(WebCore::makeExternalImageSource):
(WebCore::createAndBindTempBuffer):
(WebCore::toIntSize):
(WebCore::WebXROpaqueFramebuffer::~WebXROpaqueFramebuffer):
(WebCore::WebXROpaqueFramebuffer::startFrame):
(WebCore::WebXROpaqueFramebuffer::endFrame):
(WebCore::WebXROpaqueFramebuffer::releaseAllDisplayAttachments):
(WebCore::WebXROpaqueFramebuffer::resolveMSAAFramebuffer):
(WebCore::WebXROpaqueFramebuffer::blitShared):
(WebCore::WebXROpaqueFramebuffer::blitSharedToLayered):
(WebCore::WebXROpaqueFramebuffer::drawFramebufferSize const):
(WebCore::WebXROpaqueFramebuffer::drawViewport const):
(WebCore::displayLayout):
(WebCore::calcFramebufferPhysicalSize):
(WebCore::WebXROpaqueFramebuffer::setupFramebuffer):
(WebCore::WebXROpaqueFramebuffer::reusableDisplayAttachments const):
(WebCore::WebXROpaqueFramebuffer::bindCompositorTexturesForDisplay):
(WebCore::WebXROpaqueFramebuffer::reusableDisplayAttachmentsAtIndex):
(WebCore::WebXROpaqueFramebuffer::releaseDisplayAttachmentsAtIndex):
(WebCore::WebXROpaqueFramebuffer::bindResolveAttachments):
(WebCore::WebXROpaqueFramebuffer::calculateViewportShared): Deleted.
* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.h:
* Source/WebCore/platform/graphics/GraphicsContextGL.h:
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::blitFramebuffer):
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLTextureMapperGBM.cpp:
(WebCore::GraphicsContextGLTextureMapperGBM::createDrawingBuffer const):
(WebCore::GraphicsContextGLTextureMapperGBM::createExternalImage):
(WebCore::GraphicsContextGLTextureMapperGBM::bindExternalImage):
(WebCore::GraphicsContextGLTextureMapperGBM::enableRequiredWebXRExtensions):
* Source/WebCore/platform/graphics/gbm/GraphicsContextGLTextureMapperGBM.h:
* Source/WebCore/platform/xr/PlatformXR.h:
* Source/WebCore/testing/WebFakeXRDevice.cpp:
(WebCore::SimulatedXRDevice::initializeTrackingAndRendering):
(WebCore::SimulatedXRDevice::shutDownTrackingAndRendering):
(WebCore::SimulatedXRDevice::frameTimerFired):
(WebCore::SimulatedXRDevice::createLayerProjection):
(WebCore::SimulatedXRDevice::deleteLayer):
* Source/WebCore/testing/WebFakeXRDevice.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Shared/XR/PlatformXR.serialization.in:
* Source/WebKit/UIProcess/XR/PlatformXRSystem.cpp:
(WebKit::PlatformXRSystem::invalidate):
* Source/WebKit/UIProcess/XR/openxr/OpenXRLayer.cpp:
(WebKit::OpenXRLayer::OpenXRLayer):
(WebKit::OpenXRLayer::exportOpenXRTexture):
(WebKit::OpenXRLayerProjection::create):
(WebKit::OpenXRLayerProjection::OpenXRLayerProjection):
(WebKit::OpenXRLayerProjection::startFrame):
* Source/WebKit/UIProcess/XR/openxr/OpenXRLayer.h:
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
(WebKit::OpenXRCoordinator::collectSwapchainFormatsIfNeeded):
(WebKit::OpenXRCoordinator::createSwapchain):
(WebKit::OpenXRCoordinator::createLayerProjection):
(WebKit::OpenXRCoordinator::scheduleAnimationFrame):
(WebKit::OpenXRCoordinator::populateFrameData):
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.h:

Canonical link: https://commits.webkit.org/298184@main
